### PR TITLE
ARROW-8937: [C++] Implement strptime scalar string to timestamp kernel

### PR DESCRIPTION
**Pull request by wesm**
For now, this is not exposed in any public APIs beyond using `CallFunction` to invoke it.

### CODE REVIEW

File: cpp/src/arrow/compute/test_util.h [hunk 1]
```diff
@@ -82,6 +83,26 @@ struct DatumEqual<Type, enable_if_integer<Type>> {
   }
 };
 
+static inline void CheckScalarUnary(std::string func_name,
```
Comment (pitrou) [inline cpp/src/arrow/compute/test_util.h:86]: Shouldn't this be put in a `test_util.cc`?

Comment (wesm) [reply cpp/src/arrow/compute/test_util.h:86]: Yes it should, will do

**Comment by wesm** [comment]
+1. Will merge once CI is green

**Comment by wesm** [comment]
The ARM failure looks like a flake

```
/usr/bin/ccache /usr/bin/c++  -DARROW_EXPORTING -DARROW_HAVE_NEON -DARROW_HDFS -DARROW_JEMALLOC -DARROW_JEMALLOC_INCLUDE_DIR="" -DARROW_NO_DEPRECATED_API -DARROW_USE_GLOG -DARROW_WITH_BACKTRACE -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -DURI_STATIC_BUILD -Isrc -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /arrow/cpp/thirdparty/flatbuffers/include -isystem jemalloc_ep-prefix/src -isystem googletest_ep-prefix/include -isystem /arrow/cpp/thirdparty/hadoop/include -isystem orc_ep-install/include -Wno-noexcept-type  -fdiagnostics-color=always -ggdb -O0  -Wall -Wno-conversion -Wno-sign-conversion -Wno-unused-variable -Werror -march=armv8-a  -g -fPIC   -std=gnu++11 -MD -MT src/arrow/CMakeFiles/arrow_objlib.dir/compute/kernels/vector_take.cc.o -MF src/arrow/CMakeFiles/arrow_objlib.dir/compute/kernels/vector_take.cc.o.d -o src/arrow/CMakeFiles/arrow_objlib.dir/compute/kernels/vector_take.cc.o -c /arrow/cpp/src/arrow/compute/kernels/vector_take.cc
c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
```

This may be related to code size -- some near-term refactoring I am undertaking in this file should address that

**Pull request closed without merging**